File: pcces-go/internal/storage/sqlite/cost_structure.go
package sqlite

import (
	"context"
	"database/sql"
	"strings"
	"time"

	errx "github.com/liuxb99/pcces-web/pcces-go/internal/platform/errors"
)

type CostStructureType struct {
	ID          string `json:"id"`
	Code        string `json:"code"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Source      string `json:"source"`
	Version     string `json:"version"`
	Enabled     bool   `json:"enabled"`
	CreatedBy   string `json:"created_by"`
	CreatedAt   string `json:"created_at"`
	UpdatedAt   string `json:"updated_at"`
	RowVersion  int64  `json:"row_version"`
}

type ProjectCostStructure struct {
	ProjectCode         string `json:"project_code"`
	CostStructureTypeID string `json:"cost_structure_type_id"`
	Issue               string `json:"issue"`
	AssignedBy          string `json:"assigned_by"`
	AssignedAt          string `json:"assigned_at"`
	RowVersion          int64  `json:"row_version"`
	TypeCode            string `json:"type_code"`
	TypeName            string `json:"type_name"`
	TypeVersion         string `json:"type_version"`
	DeepLink            string `json:"deep_link"`
}

type CostStructureRepository struct {
	store *Store
}

func NewCostStructureRepository(store *Store) *CostStructureRepository {
	return &CostStructureRepository{store: store}
}

func (r *CostStructureRepository) ListTypes(ctx context.Context, enabledOnly bool) ([]CostStructureType, error) {
	query := `SELECT id, code, name, description, source, version, enabled, created_by, created_at, updated_at, row_version FROM cost_structure_types`
	if enabledOnly {
		query += ` WHERE enabled = 1`
	}
	query += ` ORDER BY code`

	rows, err := r.store.db.QueryContext(ctx, query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	items := []CostStructureType{}
	for rows.Next() {
		var item CostStructureType
		var enabled int
		if err := rows.Scan(
			&item.ID, &item.Code, &item.Name, &item.Description, &item.Source,
			&item.Version, &enabled, &item.CreatedBy, &item.CreatedAt,
			&item.UpdatedAt, &item.RowVersion,
		); err != nil {
			return nil, err
		}
		item.Enabled = enabled != 0
		items = append(items, item)
	}
	return items, rows.Err()
}

func (r *CostStructureRepository) SaveType(ctx context.Context, item CostStructureType) (CostStructureType, error) {
	item.ID = strings.TrimSpace(item.ID)
	item.Code = strings.ToUpper(strings.TrimSpace(item.Code))
	item.Name = strings.TrimSpace(item.Name)
	if item.ID == "" || item.Code == "" || item.Name == "" {
		return CostStructureType{}, errx.New(errx.CodeInvalidArgument, "id, code and name are required", "P4-COST-001")
	}
	if item.Source == "" {
		item.Source = "LEGACY"
	}
	item.Source = strings.ToUpper(strings.TrimSpace(item.Source))
	if item.Version == "" {
		item.Version = "1"
	}

	now := time.Now().UTC().Format(time.RFC3339Nano)
	tx, err := r.store.db.BeginTx(ctx, nil)
	if err != nil {
		return CostStructureType{}, err
	}
	defer tx.Rollback()

	var current int64
	err = tx.QueryRowContext(ctx, `SELECT row_version FROM cost_structure_types WHERE id = ?`, item.ID).Scan(&current)
	switch {
	case err == sql.ErrNoRows:
		if item.RowVersion != 0 {
			return CostStructureType{}, errx.New(errx.CodeConflict, "stale row_version", "P4-COST-001")
		}
		_, err = tx.ExecContext(ctx, `
			INSERT INTO cost_structure_types
			(id, code, name, description, source, version, enabled, created_by, created_at, updated_at, row_version)
			VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1)`,
			item.ID, item.Code, item.Name, item.Description, item.Source, item.Version,
			costStructureBoolInt(item.Enabled), item.CreatedBy, now, now,
		)
	case err != nil:
		return CostStructureType{}, err
	default:
		if current != item.RowVersion {
			return CostStructureType{}, errx.New(errx.CodeConflict, "stale row_version", "P4-COST-001")
		}
		var result sql.Result
		result, err = tx.ExecContext(ctx, `
			UPDATE cost_structure_types
			SET code = ?, name = ?, description = ?, source = ?, version = ?, enabled = ?, updated_at = ?, row_version = row_version + 1
			WHERE id = ? AND row_version = ?`,
			item.Code, item.Name, item.Description, item.Source, item.Version,
			costStructureBoolInt(item.Enabled), now, item.ID, item.RowVersion,
		)
		if err == nil {
			rowsAffected, rowsErr := result.RowsAffected()
			if rowsErr != nil {
				return CostStructureType{}, rowsErr
			}
			if rowsAffected != 1 {
				return CostStructureType{}, errx.New(errx.CodeConflict, "stale row_version", "P4-COST-001")
			}
		}
	}
	if err != nil {
		return CostStructureType{}, err
	}
	if err := tx.Commit(); err != nil {
		return CostStructureType{}, err
	}
	return r.GetType(ctx, item.ID)
}

func (r *CostStructureRepository) GetType(ctx context.Context, id string) (CostStructureType, error) {
	var item CostStructureType
	var enabled int
	err := r.store.db.QueryRowContext(ctx, `
		SELECT id, code, name, description, source, version, enabled, created_by, created_at, updated_at, row_version
		FROM cost_structure_types WHERE id = ?`, id,
	).Scan(
		&item.ID, &item.Code, &item.Name, &item.Description, &item.Source,
		&item.Version, &enabled, &item.CreatedBy, &item.CreatedAt,
		&item.UpdatedAt, &item.RowVersion,
	)
	if err == sql.ErrNoRows {
		return item, errx.New(errx.CodeNotFound, "cost structure type not found", "P4-COST-001")
	}
	if err != nil {
		return item, err
	}
	item.Enabled = enabled != 0
	return item, nil
}

func (r *CostStructureRepository) AssignProject(
	ctx context.Context,
	projectCode string,
	typeID string,
	issue string,
	actor string,
	rowVersion int64,
) (ProjectCostStructure, error) {
	projectCode = strings.TrimSpace(projectCode)
	typeID = strings.TrimSpace(typeID)
	issue = strings.ToUpper(strings.TrimSpace(issue))
	if issue == "" {
		issue = "BUD"
	}
	if projectCode == "" || typeID == "" {
		return ProjectCostStructure{}, errx.New(errx.CodeInvalidArgument, "project_code and cost_structure_type_id are required", "P4-COST-003")
	}
	if issue != "BUD" && issue != "BID" {
		return ProjectCostStructure{}, errx.New(errx.CodeInvalidArgument, "issue must be BUD or BID", "P4-COST-003")
	}

	tx, err := r.store.db.BeginTx(ctx, nil)
	if err != nil {
		return ProjectCostStructure{}, err
	}
	defer tx.Rollback()

	var enabled int
	err = tx.QueryRowContext(ctx, `SELECT enabled FROM cost_structure_types WHERE id = ?`, typeID).Scan(&enabled)
	if err == sql.ErrNoRows || enabled == 0 {
		return ProjectCostStructure{}, errx.New(errx.CodeNotFound, "enabled cost structure type not found", "P4-COST-003")
	}
	if err != nil {
		return ProjectCostStructure{}, err
	}

	var current int64
	err = tx.QueryRowContext(ctx, `SELECT row_version FROM project_cost_structures WHERE project_code = ?`, projectCode).Scan(&current)
	now := time.Now().UTC().Format(time.RFC3339Nano)
	switch {
	case err == sql.ErrNoRows:
		if rowVersion != 0 {
			return ProjectCostStructure{}, errx.New(errx.CodeConflict, "stale row_version", "P4-COST-003")
		}
		_, err = tx.ExecContext(ctx, `
			INSERT INTO project_cost_structures
			(project_code, cost_structure_type_id, issue, assigned_by, assigned_at, row_version)
			VALUES (?, ?, ?, ?, ?, 1)`, projectCode, typeID, issue, actor, now)
	case err != nil:
		return ProjectCostStructure{}, err
	default:
		if current != rowVersion {
			return ProjectCostStructure{}, errx.New(errx.CodeConflict, "stale row_version", "P4-COST-003")
		}
		var result sql.Result
		result, err = tx.ExecContext(ctx, `
			UPDATE project_cost_structures
			SET cost_structure_type_id = ?, issue = ?, assigned_by = ?, assigned_at = ?, row_version = row_version + 1
			WHERE project_code = ? AND row_version = ?`, typeID, issue, actor, now, projectCode, rowVersion)
		if err == nil {
			rowsAffected, rowsErr := result.RowsAffected()
			if rowsErr != nil {
				return ProjectCostStructure{}, rowsErr
			}
			if rowsAffected != 1 {
				return ProjectCostStructure{}, errx.New(errx.CodeConflict, "stale row_version", "P4-COST-003")
			}
		}
	}
	if err != nil {
		return ProjectCostStructure{}, err
	}
	if err := tx.Commit(); err != nil {
		return ProjectCostStructure{}, err
	}
	return r.GetProject(ctx, projectCode)
}

func (r *CostStructureRepository) GetProject(ctx context.Context, projectCode string) (ProjectCostStructure, error) {
	var item ProjectCostStructure
	err := r.store.db.QueryRowContext(ctx, `
		SELECT p.project_code, p.cost_structure_type_id, p.issue, p.assigned_by, p.assigned_at,
		       p.row_version, t.code, t.name, t.version
		FROM project_cost_structures p
		JOIN cost_structure_types t ON t.id = p.cost_structure_type_id
		WHERE p.project_code = ?`, projectCode,
	).Scan(
		&item.ProjectCode, &item.CostStructureTypeID, &item.Issue, &item.AssignedBy,
		&item.AssignedAt, &item.RowVersion, &item.TypeCode, &item.TypeName, &item.TypeVersion,
	)
	if err == sql.ErrNoRows {
		return item, errx.New(errx.CodeNotFound, "project cost structure not found", "P4-COST-003")
	}
	if err != nil {
		return item, err
	}
	item.DeepLink = "/app/cost-structure?project=" + projectCode
	return item, nil
}

func costStructureBoolInt(value bool) int {
	if value {
		return 1
	}
	return 0
}

File: api/cost_structure.py
"""Phase 4 cost-structure catalog and project assignment.

Legacy anchors: CostStructureTypePicker, CostStructureImport,
FormBudgetCostStructurePicker and FormBudgetCostProperty.
"""
from __future__ import annotations

from datetime import datetime, timezone
from uuid import uuid4

from flask import Blueprint, jsonify, request
from sqlalchemy import Boolean, Column, DateTime, Integer, MetaData, String, Table, Text, and_, select

metadata = MetaData()

cost_structure_types = Table(
    "cost_structure_types", metadata,
    Column("id", String(100), primary_key=True),
    Column("code", String(100), nullable=False, unique=True, index=True),
    Column("name", String(300), nullable=False),
    Column("description", Text, nullable=False, default=""),
    Column("source", String(100), nullable=False),
    Column("version", String(50), nullable=False),
    Column("enabled", Boolean, nullable=False, default=True),
    Column("created_by", String(100), nullable=False),
    Column("created_at", DateTime(timezone=True), nullable=False),
    Column("updated_at", DateTime(timezone=True), nullable=False),
    Column("row_version", Integer, nullable=False, default=1),
)

project_cost_structures = Table(
    "project_cost_structures", metadata,
    Column("project_code", String(100), primary_key=True),
    Column("cost_structure_type_id", String(100), nullable=False, index=True),
    Column("issue", String(100), nullable=False, default="BUD"),
    Column("assigned_by", String(100), nullable=False),
    Column("assigned_at", DateTime(timezone=True), nullable=False),
    Column("row_version", Integer, nullable=False, default=1),
)


class CostStructureService:
    def __init__(self, engine):
        self.engine = engine
        metadata.create_all(engine)

    @staticmethod
    def _serialize(row) -> dict:
        item = dict(row)
        for key in ("created_at", "updated_at", "assigned_at"):
            if item.get(key) is not None:
                item[key] = item[key].isoformat()
        return item

    def list_types(self, enabled_only: bool = True) -> list[dict]:
        stmt = select(cost_structure_types)
        if enabled_only:
            stmt = stmt.where(cost_structure_types.c.enabled.is_(True))
        stmt = stmt.order_by(cost_structure_types.c.code)
        with self.engine.connect() as conn:
            return [self._serialize(row) for row in conn.execute(stmt).mappings().all()]

    def save_type(self, type_id: str, body: dict, actor: str) -> dict:
        code = str(body.get("code", "")).strip().upper()
        name = str(body.get("name", "")).strip()
        if not code or not name:
            raise ValueError("code and name are required")
        now = datetime.now(timezone.utc)
        with self.engine.begin() as conn:
            current = conn.execute(select(cost_structure_types).where(cost_structure_types.c.id == type_id)).mappings().first()
            requested = int(body.get("row_version", 0))
            values = dict(
                code=code, name=name,
                description=str(body.get("description", current["description"] if current else "")),
                source=str(body.get("source", current["source"] if current else "LEGACY")).strip().upper(),
                version=str(body.get("version", current["version"] if current else "1")),
                enabled=bool(body.get("enabled", current["enabled"] if current else True)),
                updated_at=now,
            )
            if current:
                if requested != int(current["row_version"]):
                    raise RuntimeError("CONFLICT")
                result = conn.execute(cost_structure_types.update().where(and_(
                    cost_structure_types.c.id == type_id,
                    cost_structure_types.c.row_version == requested,
                )).values(**values, row_version=requested + 1))
                if result.rowcount != 1:
                    raise RuntimeError("CONFLICT")
            else:
                conn.execute(cost_structure_types.insert().values(
                    id=type_id or str(uuid4()), created_by=actor, created_at=now,
                    row_version=1, **values,
                ))
        return self.get_type(type_id)

    def get_type(self, type_id: str) -> dict:
        with self.engine.connect() as conn:
            row = conn.execute(select(cost_structure_types).where(cost_structure_types.c.id == type_id)).mappings().first()
        if not row:
            raise LookupError("cost structure type not found")
        return self._serialize(row)

    def assign_project(self, project_code: str, type_id: str, issue: str, row_version: int, actor: str) -> dict:
        project_code = project_code.strip()
        issue = issue.strip().upper() or "BUD"
        if not project_code:
            raise ValueError("project_code is required")
        if issue not in {"BUD", "BID"}:
            raise ValueError("issue must be BUD or BID")
        now = datetime.now(timezone.utc)
        with self.engine.begin() as conn:
            structure = conn.execute(select(cost_structure_types).where(and_(
                cost_structure_types.c.id == type_id,
                cost_structure_types.c.enabled.is_(True),
            ))).mappings().first()
            if not structure:
                raise LookupError("enabled cost structure type not found")
            current = conn.execute(select(project_cost_structures).where(project_cost_structures.c.project_code == project_code)).mappings().first()
            if current:
                if int(current["row_version"]) != int(row_version):
                    raise RuntimeError("CONFLICT")
                conn.execute(project_cost_structures.update().where(and_(
                    project_cost_structures.c.project_code == project_code,
                    project_cost_structures.c.row_version == row_version,
                )).values(cost_structure_type_id=type_id, issue=issue, assigned_by=actor,
                         assigned_at=now, row_version=row_version + 1))
            else:
                if int(row_version) != 0:
                    raise RuntimeError("CONFLICT")
                conn.execute(project_cost_structures.insert().values(
                    project_code=project_code, cost_structure_type_id=type_id, issue=issue,
                    assigned_by=actor, assigned_at=now, row_version=1,
                ))
        return self.get_project(project_code)

    def get_project(self, project_code: str) -> dict:
        stmt = select(project_cost_structures, cost_structure_types.c.code.label("type_code"),
                      cost_structure_types.c.name.label("type_name"), cost_structure_types.c.version.label("type_version"))\
            .join(cost_structure_types, cost_structure_types.c.id == project_cost_structures.c.cost_structure_type_id)\
            .where(project_cost_structures.c.project_code == project_code)
        with self.engine.connect() as conn:
            row = conn.execute(stmt).mappings().first()
        if not row:
            raise LookupError("project cost structure not found")
        item = self._serialize(row)
        item["deep_link"] = f"/app/cost-structure?project={project_code}"
        return item


def build_cost_structure_blueprint(service: CostStructureService, resolve_user_id):
    bp = Blueprint("cost_structure", __name__, url_prefix="/api/cost-structures")

    def actor() -> str:
        value = resolve_user_id()
        if value is None:
            raise PermissionError("authentication required")
        return str(value)

    @bp.get("/types")
    def list_types():
        enabled_only = request.args.get("enabled_only", "true").lower() not in {"false", "0", "no"}
        return jsonify(service.list_types(enabled_only))

    @bp.put("/types/<type_id>")
    def save_type(type_id: str):
        try:
            return jsonify(service.save_type(type_id, request.get_json(silent=True) or {}, actor()))
        except PermissionError as exc:
            return jsonify({"code": "UNAUTHORIZED", "detail": str(exc)}), 401
        except ValueError as exc:
            return jsonify({"code": "INVALID_ARGUMENT", "detail": str(exc)}), 400
        except RuntimeError:
            return jsonify({"code": "CONFLICT", "detail": "stale row_version"}), 409

    @bp.get("/projects/<project_code>")
    def get_project(project_code: str):
        try:
            return jsonify(service.get_project(project_code))
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404

    @bp.put("/projects/<project_code>")
    def assign_project(project_code: str):
        body = request.get_json(silent=True) or {}
        try:
            return jsonify(service.assign_project(project_code, str(body.get("cost_structure_type_id", "")),
                                                  str(body.get("issue", "BUD")), int(body.get("row_version", 0)), actor()))
        except PermissionError as exc:
            return jsonify({"code": "UNAUTHORIZED", "detail": str(exc)}), 401
        except LookupError as exc:
            return jsonify({"code": "NOT_FOUND", "detail": str(exc)}), 404
        except ValueError as exc:
            return jsonify({"code": "INVALID_ARGUMENT", "detail": str(exc)}), 400
        except RuntimeError:
            return jsonify({"code": "CONFLICT", "detail": "stale row_version"}), 409

    return bp
